src/cluster/src/check.rs

```diff
@@ -135,33 +150,19 @@ pub enum CheckError {
     /// Error while fetching create permissions for a resource
     #[error("Unable to fetch permissions")]
     FetchPermissionError,
-
-    /// One or more pre flight checks have failed
-    #[error("Some pre-install checks have failed.")]
-    PreFlightCheckError,
-}
-
-/// Captures the status of the check
-pub(crate) enum StatusCheck {
-    /// Everything seems to be working as expected, check passed
-    Working(String),
-    /// Check failed due to an error, there is no work around
-    NotWorkingNoRemediation(CheckError),
-    /// Check failed due to an error, there is a work around fix
-    NotWorking(CheckError, String),
 }
 
 #[async_trait]
 pub(crate) trait InstallCheck: Send + Sync + 'static {
     /// perform check, if successful return success message, if fail, return fail message
-    async fn perform_check(&self) -> Result<StatusCheck, CheckError>;
+    async fn perform_check(&self) -> Result<String, CheckError>;
 }
 
 pub(crate) struct LoadableConfig;
 
 #[async_trait]
 impl InstallCheck for LoadableConfig {
-    async fn perform_check(&self) -> Result<StatusCheck, CheckError> {
+    async fn perform_check(&self) -> Result<String, CheckError> {
```

Comment:
Do you just disagree with the name here? The data contained in this new structure is essentially the same as what was in the older structure, only now it is more organized across the `Result` boundary where `Ok` represents "the good thing that happened" and `Err` represents "the bad thing that happened".

Previously, we had these possible variants for a `Result<StatusCheck, CheckError>`:

```rust
match result {
    Ok(StatusCheck::Working(String)) => {},
    Ok(StatusCheck::NotWorkingNoRemediation(CheckError)) => {},
    Ok(StatusCheck::NotWorking(CheckError, String)) => {},
    Err(CheckError) => {}
}
```

One problem I had with this is that it is unclear which variant a particular `CheckError` should belong to. If I encounter an error during a check, there are three possible places that error might go, and the semantic meaning of it changes depending on where it's located, meaning that it's very easy to accidentally signal the wrong meaning.

The way things are laid out now is like this:

```rust
match result {
    Ok(String) => { /* String is a success message */ }
    Err(CheckError::AutoRecoverable(RecoverableCheck)) => {},
    Err(CheckError::Unrecoverable(UnrecoverableCheck)) => {},
    Err(CheckError::AlreadyInstalled) => {},
}
```

I specifically divided all of the variants into separate types, `RecoverableCheck` and `UnrecoverableCheck` to communicate the types of actions which may be taken next.

This new structure maps perfectly onto the actions that we want to take in each case.

- With `Ok(message)`, we just print the successful message
- With a `RecoverableCheck`, we want to attempt to automatically fix the problem
- With an `UnrecoverableCheck`, we know we can't do anything about it and we must return
- With `AlreadyInstalled`, we simply don't have any work to do

I really like to think of this less in terms of handling "errors", but rather to think about this as just a data structure that is communicating the current state of events. In our pre-checking, we have a handful of Checks we want to make, and we want to know a summary of which checks succeeded and which ones did not. The new `CheckResults` type is just a `Vec<Result<String, CheckError>>` which holds the results of _all_ of the checks. The important thing about this is that a single `CheckResults` is a perfect summary of all of the checks that have been run. Because of this, all of the command-line printing code was able to be extracted in to the `cli` module of the crate which is only compiled with the `features = ["cli"]` flag, minimizing our library's size and stdout-side-effects. Printing the status of the checks then just becomes a simple iteration over the `CheckResults`.

If you look, the `fluvio cluster check` command never actually "bubbles up" anything contained in this `CheckResults`, it is strictly used as a data structure that is iterated over and printed. In the `fluvio cluster start` command, it is also not "bubbled up". The `CheckResults` are now returned as a piece of data inside a new return struct `StartStatus` which is returned in the `Ok()` variant of the `ClusterInstaller::install_fluvio` and `LocalClusterInstaller::install` functions. It is just a description of the summary of the checks, and it is never bubbled up as an error, just looked at as a piece of data. 

I really think that the changes here are a big improvement. I think everything is laid out in a way that makes sense and works well and can be checked off and we can move on to the next thing.

Reply:
Before, it was clear what was the purpose of `performCheck`.   It was supposed perform validation to ensure Cluster meet all requirements before starting.  It's tempting to use Error for everything since it make easier to consolidate it but in this case, it is not semantically correctly.   For example, if I check if status of the network and network is down, that is not error because  that is what code is supposed to handle.   What could be considered as error?  For example if network driver is not installed because that's not what supposed to happen.   Without clear separation, it will be hard to maintain the code